core/src/retransmit_stage.rs

```diff
@@ -235,49 +235,75 @@ fn check_if_already_received(
                 None
             } else {
                 sent.push(hash);
-                Some(shred.0)
+                Some((shred.0, shred.2))
             }
         }
         None => {
             let hash = hasher.hash_packet(packet);
             cache.put(shred, vec![hash]);
-            Some(shred.0)
+            Some((shred.0, shred.2))
         }
     }
 }
 
-fn notify_first_shred_received(
+#[derive(Default)]
+struct FirstShredRecord {
+    data_received: bool,
+    coding_received: bool,
+}
+
+fn check_if_first_shred_received(
     shred_slot: Slot,
-    rpc_subscriptions: &RpcSubscriptions,
-    sent_received_slot_notification: &Mutex<BTreeSet<Slot>>,
+    is_data_shred: bool,
+    first_shreds_received: &Mutex<BTreeMap<Slot, FirstShredRecord>>,
     root_bank: &Bank,
-) {
-    let notify_slot = {
-        let mut sent_received_slot_notification_locked =
-            sent_received_slot_notification.lock().unwrap();
-        if !sent_received_slot_notification_locked.contains(&shred_slot)
-            && shred_slot > root_bank.slot()
-        {
-            sent_received_slot_notification_locked.insert(shred_slot);
-            if sent_received_slot_notification_locked.len() > 100 {
-                let mut slots_before_root =
-                    sent_received_slot_notification_locked.split_off(&(root_bank.slot() + 1));
-                // `slots_before_root` now contains all slots <= root
-                std::mem::swap(
-                    &mut slots_before_root,
-                    &mut sent_received_slot_notification_locked,
-                );
-            }
-            Some(shred_slot)
-        } else {
-            None
+) -> bool {
+    if shred_slot <= root_bank.slot() {
+        return false;
+    }
+
+    let mut first_shreds_received_locked = first_shreds_received.lock().unwrap();
+    if !first_shreds_received_locked.contains_key(&shred_slot) {
+        info!("First time receiving a shred from slot: {}", shred_slot);
```

Comment:
Could we actually make this a datapoint, I think it would be useful for debugging purposes 😃